Vpn.Service/Downloader.cs

```diff
@@ -0,0 +1,334 @@
+{
+    public static AuthenticodeDownloadValidator Coder => new("Coder Technologies Inc.");
+
+    public async Task ValidateAsync(string path, CancellationToken ct = default)
+    {
+        FileSignatureInfo fileSigInfo;
+        await using (var fileStream = File.OpenRead(path))
+        {
+            fileSigInfo = FileSignatureInfo.GetFromFileStream(fileStream);
+        }
+
+        if (fileSigInfo.State != SignatureState.SignedAndTrusted)
+            throw new Exception(
+                $"File is not signed and trusted with an Authenticode signature: State={fileSigInfo.State}");
+
+        // Coder will only use embedded signatures because we are downloading
+        // individual binaries and not installers which can ship catalog files.
+        if (fileSigInfo.Kind != SignatureKind.Embedded)
+            throw new Exception($"File is not signed with an embedded Authenticode signature: Kind={fileSigInfo.Kind}");
+
+        // TODO: check that it's an extended validation certificate
+
+        var actualName = fileSigInfo.SigningCertificate.GetNameInfo(X509NameType.SimpleName, false);
+        if (actualName != expectedName)
+            throw new Exception(
+                $"File is signed by an unexpected certificate: ExpectedName='{expectedName}', ActualName='{actualName}'");
+    }
+}
+
+public class AssemblyVersionDownloadValidator(string expectedAssemblyVersion) : IDownloadValidator
+{
+    public Task ValidateAsync(string path, CancellationToken ct = default)
+    {
+        var info = FileVersionInfo.GetVersionInfo(path);
+        if (string.IsNullOrEmpty(info.ProductVersion))
+            throw new Exception("File ProductVersion is empty or null, was the binary compiled correctly?");
+        if (info.ProductVersion != expectedAssemblyVersion)
+            throw new Exception(
+                $"File ProductVersion is '{info.ProductVersion}', but expected '{expectedAssemblyVersion}'");
+        return Task.CompletedTask;
+    }
+}
+
+/// <summary>
+///     Combines multiple download validators into a single validator. All validators will be run in order.
+/// </summary>
+/// <param name="validators">Validators to run</param>
+public class CombinationDownloadValidator(params IDownloadValidator[] validators) : IDownloadValidator
+{
+    public async Task ValidateAsync(string path, CancellationToken ct = default)
+    {
+        foreach (var validator in validators)
+            await validator.ValidateAsync(path, ct);
+    }
+}
+
+/// <summary>
+///     Handles downloading files from the internet. Downloads are performed asynchronously using DownloadTask.
+///     Single-flight is provided to avoid performing the same download multiple times.
+/// </summary>
+public class Downloader : IDownloader
+{
+    private readonly ConcurrentDictionary<string /* DestinationPath */, DownloadTask> _downloads = new();
+    private readonly ILogger<Downloader> _logger;
+
+    // ReSharper disable once ConvertToPrimaryConstructor
```

Comment:
I don't see these get used anywhere, could disable the rule

Reply:
It's not a rule but it's a suggestion which I don't want to use in this place. I might just block the online constructors in the whole solution though and fix them all